src/js/binaryen.js-post.js

```diff
@@ -4281,6 +4281,42 @@ Module['TupleExtract'] = makeExpressionWrapper({
   }
 });
 
+// Function wrapper
+
+Module['Function'] = (() => {
+  function Function(func) {
+    if (!(this instanceof Function)) {
+      if (!func) return null;
+      return new Function(func);
+    }
+    if (!func) throw Error("function reference must not be null");
+    this['func'] = func;
```

Comment:
Forgive my ignorance of JS, but what is `this` in this context? What object has `expr` and `func` fields? And what is `makeWrapperInstanceMembers` doing?

Reply:
The `this` here is an instance of the `Function` class. The `Function` wrapper needs to reference the underlying pointer to the C API function somehow, and it stores that in `Function#func`. The funny magic above makes sure that one can use both `Function(funcPtr)` and `new Function(funcPtr)` to wrap an underlying function by its C API pointer. Similarly, the existing specific `Expression` wrappers for `Nop`, `LocalSet` etc. have an `Expression#expr` to store that reference.

`makeWrapperInstanceMembers` takes the static methods of a class, for example `Function.hasLocalName(func, index)` and transforms these to instance methods and, if applicable, properties with the first argument bound to `this['func']` for example, so a user has the option to use the API as plain functions

```js
Function.hasLocalName(func, index)
```

or *wrap* a C API function pointer in a `Function` instance for convenience

```js
var theFunc = Function(func)
theFunc.numLocalNames // property made of Function.getNumLocalNames
theFunc.hasLocalName(index) // instance method made from Function.hasLocalName
```

The latter is mostly a convenience feature for us, so we can declare the API once, and JS takes care of the rest without duplicating any code while simultaneously providing a more JS-y API for those who prefer. Does that make sense? :)

Reply:
Thanks, the goal here makes sense, but the code still seems like black magic to me. But that shouldn't stop this from landing!